File: AF_Backend/app/schemas/contribution.py
from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from decimal import Decimal
from typing import Optional, List

class ContributionBase(BaseModel):
    campaign_id: UUID
    amount: Decimal = Field(..., gt=0)

class ContributionCreate(ContributionBase):
    pass

class ContributionOut(BaseModel):
    contribution_id: UUID
    campaign_id: UUID
    contributor_id: UUID
    amount: Decimal
    status: str
    created_at: datetime

    class Config:
        from_attributes = True

class ContributionResponse(BaseModel):
    contribution: ContributionOut
    campaign_total_raised: Decimal
    escrow_balance: Decimal
    vote_token_id: Optional[UUID] = None

class UserContributionOut(BaseModel):
    contribution_id: UUID
    campaign_id: UUID
    campaign_title: str
    campaign_status: str
    amount: Decimal
    status: str
    created_at: datetime

    class Config:
        from_attributes = True

class ContributorStats(BaseModel):
    total_portfolio_value: Decimal
    active_investments_count: int

class WalletLedgerEntry(BaseModel):
    id: UUID
    campaign_title: str
    amount: Decimal
    type: str # 'contribution' or 'disbursement' (future)
    status: str
    date: datetime

class ContributorWalletStats(BaseModel):
    available_funds: Decimal
    invested_funds: Decimal
    ledger: List[WalletLedgerEntry]

    class Config:
        from_attributes = True

File: AF_FrontEnd/front_end/lib/data/models/contribution.dart
class UserContribution {
  final String contributionId;
  final String campaignId;
  final String campaignTitle;
  final String campaignStatus;
  final double amount;
  final String status;
  final DateTime createdAt;

  UserContribution({
    required this.contributionId,
    required this.campaignId,
    required this.campaignTitle,
    required this.campaignStatus,
    required this.amount,
    required this.status,
    required this.createdAt,
  });

  factory UserContribution.fromJson(Map<String, dynamic> json) {
    return UserContribution(
      contributionId: json['contribution_id'],
      campaignId: json['campaign_id'],
      campaignTitle: json['campaign_title'],
      campaignStatus: json['campaign_status'],
      amount: _parseDouble(json['amount']),
      status: json['status'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }
}
